### ISSUE TITLE
Do not add an empty cell when an entire Markdown cell gets extracted into a virtual document

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#15870]
## Description

JupyterLab 4 added support for Markdown documents by using extractors mechanisms; it seems that markdown cells are erroneously included as empty cells in the resulting Python document.

It seems that the logic in `TextForeignCodeExtractor` and in the `extractForeignCode()` or `appendCodeBlock()` need to be modified so that when an entire cell gets extracted no cell would be added (rather than adding an empty code cell, padded by two empty lines on each side).

## Reproduce

See jupyterlab-lsp issues:
- https://github.com/jupyter-lsp/jupyterlab-lsp/issues/1054
- https://github.com/jupyter-lsp/jupyterlab-lsp/issues/1040

## Expected behavior

When an entire cell is extracted do not add new lines before and after it.

This could be solved by special casing the logic in:

https://github.com/jupyterlab/jupyterlab/blob/2d681271b159ac533df1ddb576ddcb256c44e59c/packages/lsp/src/virtual/document.ts#L614-L673

For example if `lines` is just one empty line we could skip adding `this.blankLinesBetweenCells`. This need to be conditional on `foreignDocumentsMap` being non-empty so that it does not lead to false negatives (e.g. undetected empty code cells),

## Context 

- JupyterLab version: 4.1.2

**Comment by andrii-i** [comment]
@krassowski both `lsp` package vendored by jupyterlab and `jupyterlab-lsp` package implement `VirtualDocument.appendCodeBlock`. So when `jupyterlab-lsp` is installed which is necessary to get syntax highlighting and other UI features `VirtualDocument.appendCodeBlock` from `jupyterlab-lsp` is used. 

Then the same changes as in #17363 should be made in jupyterlab-lsp's `VirtualDocument.appendCodeBlock`, right? jupyterlab-lsp VirtualDocument implementation has additional logic to set metadata to skip checking certain lines which seems meaningful. 

* JupyterLab packages/lsp: https://github.com/jupyterlab/jupyterlab/blob/1bceac64fd0ae3eb941eb777972730d6a6df939d/packages/lsp/src/virtual/document.ts#L605-L673

* jupyterlab-lsp: https://github.com/jupyter-lsp/jupyterlab-lsp/blob/4628ce6bfcf1910c98b6f575d0e89e4f805c7670/packages/jupyterlab-lsp/src/virtual/document.ts#L83C3-L142

**Comment by andrii-i** [comment]
@krassowski Looking at the code it seems that upstream LSP package does propagate `skipInspect` value returned by `prepareCodeBlock` https://github.com/jupyter-lsp/jupyterlab-lsp/blob/4628ce6bfcf1910c98b6f575d0e89e4f805c7670/packages/jupyterlab-lsp/src/virtual/document.ts#L103
`        skipInspect: skipInspect[i],`

While jupyterlab lsp package does not https://github.com/jupyterlab/jupyterlab/blob/1bceac64fd0ae3eb941eb777972730d6a6df939d/packages/lsp/src/virtual/document.ts#L634

So does it mean that changing line referenced above in  jupyterlab lsp package to `skipInspect: skipInspect[i]` would fix it?

**Comment by krassowski** [comment]
Yes, adding it in jupyterlab version would mean we could remove the override in the `jupyterlab-lsp` repo.
